bundles/org.openhab.binding.iaqualink/src/main/java/org/openhab/binding/iaqualink/internal/handler/IAqualinkHandler.java

```diff
@@ -0,0 +1,450 @@
+        }
+
+        this.refresh = configuration.refresh.intValue() > MIN_REFRESH_SECONDS ? configuration.refresh.intValue()
+                : MIN_REFRESH_SECONDS;
+
+        try {
+            AccountInfo accountInfo = client.login(username, password, apiKey);
+            sessionId = accountInfo.getSessionId();
+            if (sessionId == null) {
+                throw new IOException("Response from controller not valid");
+            }
+            logger.debug("SessionID {}", sessionId);
+
+            Device[] devices = client.getDevices(apiKey, accountInfo.getAuthenticationToken(), accountInfo.getId());
+
+            if (devices.length == 0) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "No registered devices found");
+                return;
+            }
+
+            if (StringUtils.isNotBlank(confSerialId)) {
+                serialNumber = confSerialId.replaceAll("[^a-zA-Z0-9]", "").toLowerCase();
+                if (!Arrays.stream(devices).anyMatch(device -> device.getSerialNumber().equals(serialNumber))) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                            "No Device for given serialId found");
+                    return;
+                }
+            } else {
+                serialNumber = devices[0].getSerialNumber();
+            }
+
+            initPolling(0);
+
+        } catch (InterruptedException | IOException e) {
+            logger.debug("Could not connect to service");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+        } catch (NotAuthorizedException e) {
+            logger.debug("Credentials not valid");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Credentials not valid");
+        }
+    }
+
+    /**
+     * Starts/Restarts polling with an initial delay. This allows changes in the poll cycle for when commands are sent
+     * and we need to poll sooner then the next refresh cycle.
+     */
+    private synchronized void initPolling(int initalDelay) {
+        clearPolling();
+        pollFuture = scheduler.scheduleWithFixedDelay(() -> {
+            try {
+                pollController();
+            } catch (Exception e) {
+                logger.debug("Exception during poll", e);
+            }
+        }, initalDelay, refresh, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Stops/clears this thing's polling future
+     */
+    private void clearPolling() {
+        ScheduledFuture<?> localFuture = pollFuture;
+        if (localFuture != null && !localFuture.isCancelled()) {
+            logger.trace("Canceling future");
+            localFuture.cancel(true);
+        }
+    }
+
+    /**
+     * Poll the controller for updates.
+     */
+    private void pollController() {
+        try {
+
+            JsonElement home = client.getHomeJson(serialNumber, sessionId);
```

Comment:
Cant you just do this also with proper mapped objects?

Reply:
there is a one to one mapping between channels names and the returned json,  by not having to completely deserialize the object i A) avoid having to write a TON of boiler plate code and 2) do not incur the performance penalties of things like reflection costs.  I much prefer to handle the response this way.

Reply:
I agree this was sloppy, i made a compromise to clean this up.